src/lib/apis/fetchForeignAPI.ts

```diff
@@ -0,0 +1,48 @@
+import fetch, { Request, RequestInfo, Response } from 'node-fetch';
+import { ZodSchema } from 'zod';
+import { Logger } from '../logger';
+
+/**
+ * Fetch data from any API endpoint that returns JSON formatted data.
+ * @typeParam ReturnType - The expected type of the returned data.
+ * @param request The [Request](https://developer.mozilla.org/en-US/docs/Web/API/Request) object to be passed to `fetch()`.
+ * @param zodSchema The [Zod](https://github.com/colinhacks/zod) schema that the returned data conforms to. The promise will reject if the returned data does not conform to the schema provided.
+ * @returns A promise that resolves to the expected type or rejects with an [Error](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error) or a {@link ZodError} if the validation failed.
+ */
+export const fetchForeignAPI = async <ReturnType = unknown>(request: RequestInfo, zodSchema: ZodSchema<ReturnType>, debug?: boolean): Promise<ReturnType> => {
+    const req = new Request(request);
+
+    let response: Response;
+    try {
+        response = await fetch(req);
+    } catch (e) {
+        throw new Error(`An error occured while fetching data from ${req.url}: ${String(e)}`);
+    }
+
+    if (!response.ok) {
+        throw new Error(`HTTP Error. Status: ${response.status}`);
+    }
+
+    let data: unknown;
+    try {
+        data = await response.json();
+    } catch (e) {
+        throw new Error(`Could not parse JSON. Make sure the endpoint at ${req.url} returns valid JSON. Error: ${String(e)}`);
+    }
+
+    const result = zodSchema.safeParse(data);
+
+    if (!result.success) {
+        Logger.error("[zod] Data validation failed! Pass the 'debug' flag to 'fetchForeignAPI()' to dump the retrieved data to the console.");
+        Logger.error(`Endpoint location: ${req.url}.`);
+        if (debug) {
+            // winston doesn't log objects in a useful way at the moment
+            // eslint-disable-next-line no-console
+            console.log('RETRIEVED DATA:', data);
```

Comment:
Does this actually print the object and not just `[Object object]`? Seems like the solution would be 
```suggestion
            Logger.debug('RETRIEVED DATA:', typeof data === 'object' ? JSON.stringify(data) : data);
```
Just not a fan of console.log being directly called, Logger allows us to centrally manage any logging configuration we'd like, starting to deviate might be hard to track.

Reply:
I totally agree with you that having `console.log()` spread across the project is less than ideal. 

However, by default, winston doesn't do anything with the `meta` object. It is up to the implementing application to use it as they wish. I agree with you that this is how the logger should work but we currently do nothing with the meta object. See [logger.ts](https://github.com/flybywiresim/discord-bot-utils/blob/staging/src/lib/logger.ts).

I think @benw202 wanted to refactor the entire logging architecture so this might be something to keep in mind for that. Using our own formatter also allows for color coded logs, etc.

And to answer your question regarding `[object Object]`, `console.log` stringifies objects correctly as long as they are passed as separate args.

Let me know if I missed something here. I tested your suggestion but for me that just outputs the string `RETRIEVED DATA`.

Reply:
Winston should handle the object just fine as well without needing to stringify it if its an object FYI. I am pretty sure it should be able to regardless of wrapping it into another object but I have not tested that.
e.g. 
```js
Logger.debug('RETRIEVED DATA', { data } )```

Reply:
There might indeed be a difference between passing the data directly vs wrapping it in an object. I'm not available to test this until the weekend but will do once I'm back.

Will report back and mark as draft until then.

Reply:
There might indeed be a difference between passing the data directly vs wrapping it in an object. I'm not available to test this until the weekend but will do once I'm back.

Will report back and mark as draft until then.

Reply:
There might indeed be a difference between passing the data directly vs wrapping it in an object. I'm not available to test this until the weekend but will do once I'm back.

Will report back and mark as draft until then.

Reply:
So I have looked into this and indeed, winston logs the meta object when using any log method except `Logger.log()`. I probably didn't read the documentation well enough to catch that mistake.

Latest commit has the console removed 👍